Cannot use downward API in `PackageBuild`

**What steps did you take:**

1. `kctrl package init`
1. Update `PackageBuild` to read values from the downward API

    ```yaml
    apiVersion: kctrl.carvel.dev/v1alpha1
    kind: PackageBuild
    metadata:
      creationTimestamp: null
      name: downward.example.com
    spec:
      release:
      - resource: {}
      template:
        spec:
          app:
            spec:
              deploy:
              - kapp: {}
              template:
              - ytt:
                  paths:
                  - carvel
                  valuesFrom:
                  - downwardAPI:
                      items:
                      - kubernetesAPIs: {}
                        name: k8s_gvs
              - kbld: {}
          export:
          - imgpkgBundle:
              useKbldImagesLock: true
            includePaths:
            - carvel
    ```

1. Release/build the package

    ```shell
    kctrl package release --version 0.0.0-mamachanko --tag 0.0.0-mamachanko --copy-to build --yes
    ```


**What happened:**

```
Prerequisites
1. Host is authorized to push images to a registry (can be set up by running
`docker login`)
2. `package init` ran successfully.

The bundle created needs to be pushed to an OCI registry. (format:
<REGISTRY_URL/REPOSITORY_NAME>) e.g. index.docker.io/k8slt/sample-bundlekbld builds images when necessary and ensures that all image references are
resolved to an immutable reference
Building images and resolving references
panic: Not implemented

goroutine 1 [running]:
github.com/vmware-tanzu/carvel-kapp-controller/cli/pkg/kctrl/local.(*MinCoreClient).Discovery(0xc000865ee0)
        github.com/vmware-tanzu/carvel-kapp-controller/cli/pkg/kctrl/local/min_core_client.go:94 +0x25
github.com/vmware-tanzu/carvel-kapp-controller/pkg/componentinfo.(*ComponentInfo).KubernetesAPIs(0x21a48c8?)
        github.com/vmware-tanzu/carvel-kapp-controller@v0.46.0/pkg/componentinfo/component_info.go:80 +0x2d
github.com/vmware-tanzu/carvel-kapp-controller/pkg/app.(*App).template.(*App).buildDownwardAPIAdditionalValues.func3()
        github.com/vmware-tanzu/carvel-kapp-controller@v0.46.0/pkg/app/app_template.go:96 +0x3f
github.com/vmware-tanzu/carvel-kapp-controller/pkg/template.DownwardAPIValues.AsYAMLs({{0xc000022480, 0x1, 0x1}, {{{0x0, 0x0}, {0x0, 0x0}}, {{0x271800b, 0xd}, {0x2713bbf, ...}, ...}}, ...})
        github.com/vmware-tanzu/carvel-kapp-controller@v0.46.0/pkg/template/downward_api_values.go:74 +0x519
github.com/vmware-tanzu/carvel-kapp-controller/pkg/template.Values.writeFromDownwardAPI({{0xc00032e540, 0x1, 0x1}, {0xc0003788d0, 0xc0003788c0, 0xc0003788e0}, {{0x271800b, 0xd}, {0x2713bbf, 0x9}, ...}, ...}, ...)
        github.com/vmware-tanzu/carvel-kapp-controller@v0.46.0/pkg/template/values.go:117 +0x58
github.com/vmware-tanzu/carvel-kapp-controller/pkg/template.Values.AsPaths({{0xc00032e540, 0x1, 0x1}, {0xc0003788d0, 0xc0003788c0, 0xc0003788e0}, {{0x271800b, 0xd}, {0x2713bbf, 0x9}, ...}, ...}, ...)
        github.com/vmware-tanzu/carvel-kapp-controller@v0.46.0/pkg/template/values.go:74 +0x51b
github.com/vmware-tanzu/carvel-kapp-controller/pkg/template.(*Ytt).template(0xc000395080, {0xc00030e070, 0x62}, {0x0?, 0x0})
        github.com/vmware-tanzu/carvel-kapp-controller@v0.46.0/pkg/template/ytt.go:66 +0x588
github.com/vmware-tanzu/carvel-kapp-controller/pkg/template.(*Ytt).TemplateDir(0xc000866a60?, {0xc00030e070?, 0xc0000061a0?})
        github.com/vmware-tanzu/carvel-kapp-controller@v0.46.0/pkg/template/ytt.go:38 +0x45
github.com/vmware-tanzu/carvel-kapp-controller/pkg/app.(*App).template(0xc00017e000, {0xc00030e070, 0x62})
        github.com/vmware-tanzu/carvel-kapp-controller@v0.46.0/pkg/app/app_template.go:53 +0x139b
github.com/vmware-tanzu/carvel-kapp-controller/pkg/app.(*App).reconcileFetchTemplateDeploy(0xc00017e000)
        github.com/vmware-tanzu/carvel-kapp-controller@v0.46.0/pkg/app/app_reconcile.go:142 +0x452
github.com/vmware-tanzu/carvel-kapp-controller/pkg/app.(*App).reconcileDeploy(0xc00017e000)
        github.com/vmware-tanzu/carvel-kapp-controller@v0.46.0/pkg/app/app_reconcile.go:93 +0x65
github.com/vmware-tanzu/carvel-kapp-controller/pkg/app.(*App).Reconcile(0xc00017e000, 0x0)
        github.com/vmware-tanzu/carvel-kapp-controller@v0.46.0/pkg/app/app_reconcile.go:50 +0x2ea
github.com/vmware-tanzu/carvel-kapp-controller/pkg/app.(*CRDApp).Reconcile(...)
        github.com/vmware-tanzu/carvel-kapp-controller@v0.46.0/pkg/app/crd_app.go:121
github.com/vmware-tanzu/carvel-kapp-controller/pkg/app.(*Reconciler).Reconcile(0xc0002e80f0, {0x2996e28, 0x3757680}, {{{0x2713bbf, 0x9}, {0x271800b, 0xd}}})
        github.com/vmware-tanzu/carvel-kapp-controller@v0.46.0/pkg/app/reconciler.go:99 +0x3f4
github.com/vmware-tanzu/carvel-kapp-controller/cli/pkg/kctrl/local.(*Reconciler).Reconcile(0xc000868fd0, {{0xc000818200, 0x1, 0x1}, {0x0, 0x0, 0x0}, {0x0, 0x0, 0x0}, ...}, ...)
        github.com/vmware-tanzu/carvel-kapp-controller/cli/pkg/kctrl/local/reconciler.go:144 +0x7e5
github.com/vmware-tanzu/carvel-kapp-controller/cli/pkg/kctrl/cmd/app/release.(*AppSpecBuilder).Build(0xc000869908)
        github.com/vmware-tanzu/carvel-kapp-controller/cli/pkg/kctrl/cmd/app/release/app_spec_builder.go:102 +0x879
github.com/vmware-tanzu/carvel-kapp-controller/cli/pkg/kctrl/cmd/package/release.(*ReleaseOptions).Run(0xc0001a5400)
        github.com/vmware-tanzu/carvel-kapp-controller/cli/pkg/kctrl/cmd/package/release/release.go:127 +0x505
github.com/vmware-tanzu/carvel-kapp-controller/cli/pkg/kctrl/cmd/package/release.NewReleaseCmd.func1(0xc000593230?, {0x27141d1?, 0xa?, 0x24cd920?})
        github.com/vmware-tanzu/carvel-kapp-controller/cli/pkg/kctrl/cmd/package/release/release.go:54 +0x17
github.com/vmware-tanzu/carvel-kapp-controller/cli/pkg/kctrl/cmd.configureGlobalFlags.func1.1(0xc000213be8?, {0xc0001ea150, 0x0, 0x7})
        github.com/vmware-tanzu/carvel-kapp-controller/cli/pkg/kctrl/cmd/kctrl.go:141 +0xc3
github.com/cppforlife/cobrautil.DisallowExtraArgs.WrapRunEForCmd.func2.1(0xc000213c90?, {0xc0001ea150, 0x0, 0x7})
        github.com/cppforlife/cobrautil@v0.0.0-20221130162803-acdfead391ef/misc.go:45 +0x62
github.com/vmware-tanzu/carvel-kapp-controller/cli/pkg/kctrl/cmd.configureGlobalFlags.WrapRunEForCmd.func3.1(0x11c4d3b?, {0xc0001ea150, 0x0, 0x7})
        github.com/cppforlife/cobrautil@v0.0.0-20221130162803-acdfead391ef/misc.go:45 +0x62
github.com/vmware-tanzu/carvel-kapp-controller/cli/pkg/kctrl/cmd.configureGlobalFlags.WrapRunEForCmd.func6.1(0x7?, {0xc0001ea150, 0x0, 0x7})
        github.com/cppforlife/cobrautil@v0.0.0-20221130162803-acdfead391ef/misc.go:45 +0x62
github.com/vmware-tanzu/carvel-kapp-controller/cli/pkg/kctrl/cmd.configureTTY.func1.WrapRunEForCmd.func2.1(0xc0000c9c00?, {0xc0001ea150, 0x0, 0x7})
        github.com/cppforlife/cobrautil@v0.0.0-20221130162803-acdfead391ef/misc.go:45 +0x62
github.com/spf13/cobra.(*Command).execute(0xc00083adc0, {0xc0001ea0e0, 0x7, 0x7})
        github.com/spf13/cobra@v1.6.1/command.go:850 +0x694
github.com/spf13/cobra.(*Command).ExecuteC(0xc000232b00)
        github.com/spf13/cobra@v1.6.1/command.go:958 +0x389
github.com/spf13/cobra.(*Command).Execute(0xc0000fea40?)
        github.com/spf13/cobra@v1.6.1/command.go:895 +0x13
main.nonExitingMain()
        github.com/vmware-tanzu/carvel-kapp-controller/cli/cmd/kctrl/kctrl.go:39 +0x145
main.main()
        github.com/vmware-tanzu/carvel-kapp-controller/cli/cmd/kctrl/kctrl.go:21 +0x13
```

It appears as if it wants to read the downward API values from an actual cluster using the discovery client.

**What did you expect:**

That it just builds the package.

**Anything else you would like to add:**

no

**Environment:**

```
❯ kctrl --version
kctrl version 0.50.2

Succeeded
```

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

--- a/pkg/template/downward_api_values.go
+++ b/pkg/template/downward_api_values.go
@@ -55,9 +55,9 @@ func (a DownwardAPIValues) AsYAMLs() ([][]byte, error) {
 				if err != nil {
 					return nil, err
 				}
-				keyValueContent, err = yaml.Marshal(map[string]string{item.Name: v})
+				keyValueContent, err = a.nestedKeyValue(item.Name, v)
 			} else {
-				keyValueContent, err = yaml.Marshal(map[string]string{item.Name: item.KubernetesVersion.Version})
+				keyValueContent, err = a.nestedKeyValue(item.Name, item.KubernetesVersion.Version)
 			}
 		case item.KappControllerVersion != nil:
 			if item.KappControllerVersion.Version == "" {
@@ -65,9 +65,9 @@ func (a DownwardAPIValues) AsYAMLs() ([][]byte, error) {
 				if err != nil {
 					return nil, err
 				}
-				keyValueContent, err = yaml.Marshal(map[string]string{item.Name: v})
+				keyValueContent, err = a.nestedKeyValue(item.Name, v)
 			} else {
-				keyValueContent, err = yaml.Marshal(map[string]string{item.Name: item.KappControllerVersion.Version})
+				keyValueContent, err = a.nestedKeyValue(item.Name, item.KappControllerVersion.Version)
 			}
 		case item.KubernetesAPIs != nil:
 			if item.KubernetesAPIs.GroupVersions == nil {
@@ -75,9 +75,9 @@ func (a DownwardAPIValues) AsYAMLs() ([][]byte, error) {
 				if err != nil {
 					return nil, err
 				}
-				keyValueContent, err = yaml.Marshal(map[string]interface{}{item.Name: v})
+				keyValueContent, err = a.nestedKeyValue(item.Name, v)
 			} else {
-				keyValueContent, err = yaml.Marshal(map[string]interface{}{item.Name: item.KubernetesAPIs.GroupVersions})
+				keyValueContent, err = a.nestedKeyValue(item.Name, item.KubernetesAPIs.GroupVersions)
 			}
 		default:
 			return nil, fmt.Errorf("Invalid downward API item given")

--- a/pkg/template/values_test.go
+++ b/pkg/template/values_test.go
@@ -252,48 +252,6 @@ func TestValues(t *testing.T) {
 			require.Error(t, err)
 			assert.ErrorContains(t, err, "Invalid field spec provided to DownwardAPI. Only single supported fields are allowed")
 		})
-
-		t.Run("return kubernetes cluster version if not supplied", func(t *testing.T) {
-			subject := subject
-			subject.ValuesFrom = []v1alpha1.AppTemplateValuesSource{{DownwardAPI: &v1alpha1.AppTemplateValuesDownwardAPI{
-				Items: []v1alpha1.AppTemplateValuesDownwardAPIItem{
-					{Name: "k8s-version", KubernetesVersion: &v1alpha1.Version{}},
-				}},
-			}}
-			subject.AdditionalValues = AdditionalDownwardAPIValues{
-				KubernetesVersion: func() (string, error) {
-					return "0.20.0", nil
-				},
-			}
-
-			paths, cleanup, err := subject.AsPaths(os.TempDir())
-			require.NoError(t, err)
-			t.Cleanup(cleanup)
-
-			require.Len(t, paths, 1)
-			assertFileContents(t, paths[0], "k8s-version: 0.20.0\n")
-		})
-
-		t.Run("return kapp-controller version", func(t *testing.T) {
-			subject := subject
-			subject.ValuesFrom = []v1alpha1.AppTemplateValuesSource{{DownwardAPI: &v1alpha1.AppTemplateValuesDownwardAPI{
-				Items: []v1alpha1.AppTemplateValuesDownwardAPIItem{
-					{Name: "kc-version", KappControllerVersion: &v1alpha1.Version{}},
-				}},
-			}}
-			subject.AdditionalValues = AdditionalDownwardAPIValues{
-				KappControllerVersion: func() (string, error) {
-					return "0.42.31337", nil
-				},
-			}
-
-			paths, cleanup, err := subject.AsPaths(os.TempDir())
-			require.NoError(t, err)
-			t.Cleanup(cleanup)
-
-			require.Len(t, paths, 1)
-			assertFileContents(t, paths[0], "kc-version: 0.42.31337\n")
-		})
 	})
 }
 

--- a/pkg/template/values_test.go
+++ b/pkg/template/values_test.go
@@ -65,22 +65,45 @@ func TestValues(t *testing.T) {
 
 		t.Run("name should allow nested key structure", func(t *testing.T) {
 			subject := subject
+			subject.AdditionalValues = AdditionalDownwardAPIValues{
+				KubernetesVersion: func() (string, error) {
+					return "test-kubernetes-version", nil
+				},
+				KubernetesAPIs: func() ([]string, error) {
+					return []string{"somegroup.example.com/someversion"}, nil
+				},
+				KappControllerVersion: func() (string, error) {
+					return "test-kapp-controller-version", nil
+				},
+			}
 			subject.ValuesFrom = []v1alpha1.AppTemplateValuesSource{{DownwardAPI: &v1alpha1.AppTemplateValuesDownwardAPI{
 				Items: []v1alpha1.AppTemplateValuesDownwardAPIItem{
 					{Name: "parent.child", FieldPath: "metadata.name"},
 					{Name: "parent.child1.child2", FieldPath: "metadata.namespace"},
 					{Name: "parent.childwith\\.dot", FieldPath: "metadata.namespace"},
+					{Name: "parent.kubernetes_version", KubernetesVersion: &v1alpha1.Version{}},
+					{Name: "parent.kubernetes_version_custom", KubernetesVersion: &v1alpha1.Version{Version: "test-kubernetes-version-custom"}},
+					{Name: "parent.kubernetes_apis", KubernetesAPIs: &v1alpha1.KubernetesAPIs{}},
+					{Name: "parent.kubernetes_apis_custom", KubernetesAPIs: &v1alpha1.KubernetesAPIs{GroupVersions: []string{"somecustomgroup.example.com/someversion"}}},
+					{Name: "parent.kapp_controller_version", KappControllerVersion: &v1alpha1.Version{}},
+					{Name: "parent.kapp_controller_version_custom", KubernetesVersion: &v1alpha1.Version{Version: "test-kapp-controller-version-custom"}},
 				}},
 			}}
 
 			paths, cleanup, err := subject.AsPaths(os.TempDir())
 			require.NoError(t, err)
-			require.Len(t, paths, 3)
+			require.Len(t, paths, 9)
 			t.Cleanup(cleanup)
 
 			assertFileContents(t, paths[0], "parent:\n  child: some-name\n")
 			assertFileContents(t, paths[1], "parent:\n  child1:\n    child2: some-namespace\n")
 			assertFileContents(t, paths[2], "parent:\n  childwith.dot: some-namespace\n")
+			assertFileContents(t, paths[3], "parent:\n  kubernetes_version: test-kubernetes-version\n")
+			assertFileContents(t, paths[4], "parent:\n  kubernetes_version_custom: test-kubernetes-version-custom\n")
+			assertFileContents(t, paths[5], "parent:\n  kubernetes_apis:\n  - somegroup.example.com/someversion\n")
+			assertFileContents(t, paths[6], "parent:\n  kubernetes_apis_custom:\n  - somecustomgroup.example.com/someversion\n")
+			assertFileContents(t, paths[7], "parent:\n  kapp_controller_version: test-kapp-controller-version\n")
+			assertFileContents(t, paths[8], "parent:\n  kapp_controller_version_custom: test-kapp-controller-version-custom\n")
 		})
 
 		t.Run("map field paths should allow subpaths", func(t *testing.T) {

